virttest/env_process.py

```diff
@@ -569,6 +569,64 @@ def preprocess(test, params, env):
     """
     error_context.context("preprocessing")
 
+    def _check_host_component_version(required_cpnt, cpnt, get_version_func):
+        """"
+        Check if a component on host satisfied the test component requirements.
+        If not, cancel the test.
+
+        :param required_cpnt: A tuple of required component.
+                              (None,"2.10.0") means the cpnt version must
+                              less then or equal to 2.10.0".
+                              ("2.1.0",None) means the cpnt version must
+                              greater then 2.1.0".
+                              ("2.1.0","2.10.0") means the cpnt version must
+                              less then or equal to 2.10.0" and greater then
+                              2.1.0".
+        :param cpnt: A raw component version string.
+        :param get_version_func: A function for getting actual version number.
+        """
+        from distutils.version import LooseVersion
+        host_component = get_version_func(cpnt)
+        required_cpnt = eval(required_cpnt)
+        func_info = get_version_func.__name__.replace("_", " ")
+        # successfully got the version string
+        if host_component:
+            host_cpnt_ver = LooseVersion(host_component)
+            # if low bound is provided
+            if required_cpnt[0]:
+                lower_bound = LooseVersion(required_cpnt[0])
+                # must greater then the lower bound
+                if host_cpnt_ver <= lower_bound:
+                    test.cancel("from host,%s %s, "
+                                "must greater then %s" %
+                                (func_info, host_cpnt_ver, lower_bound))
+            # if upper bound is provided
+            if required_cpnt[1]:
+                upper_bound = LooseVersion(required_cpnt[1])
+                # must less then or equal to the upper bound
+                if host_cpnt_ver > upper_bound:
+                    test.cancel("from host,%s %s, "
+                                "must less then or equal to %s" %
+                                (func_info, host_cpnt_ver, upper_bound))
+        else:
+            test.cancel("from host, can not%s" % func_info)
+
+    def _get_qemu_version(qmeu_version_str):
+        regex = re.compile(r'([0-9]+\.[0-9]+\.[0-9]\-[0-9]+)')
+        match = regex.search(qmeu_version_str)
+        if match:
+            return match.group(0)
+        else:
+            return None
+
+    def _get_kernel_version(kernel_version_str):
+        regex = re.compile(r'([0-9]+\.[0-9]+\.[0-9]\-[0-9]+)')
+        match = regex.search(kernel_version_str)
+        if match:
+            return match.group(0)
+        else:
+            return None
+
```

Comment:
These two functions are almost same. The reason for this is we want keep them separate for each component，even they are same.